pkg/envtest/komega/komega.go

```diff
@@ -0,0 +1,133 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package komega
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// komega has Gomega Matchers that use the controller-runtime client.
+type komega struct {
+	ctx    context.Context
+	client client.Client
+}
+
+var _ Komega = &komega{}
+
+// NewKomega creates a new instance with a given client.
+func NewKomega(c client.Client) Komega {
+	return &komega{
+		client: c,
+	}
+}
+
+// WithContext returns a copy using the given Context.
+func (k komega) WithContext(ctx context.Context) Komega {
+	k.ctx = ctx
+	return &k
+}
+
+func (k komega) WithClient(c client.Client) Komega {
+	k.client = c
+	return &k
+}
+
+// context returns the matcher context if one has been set.
+// Else it returns the context.TODO().
+func (k *komega) context() context.Context {
+	if k.ctx == nil {
+		return context.Background()
+	}
+	return k.ctx
+}
+
+// Get fetches an object until the forwarded error matches.
+func (k *komega) Get(obj client.Object) func() error {
+	key := types.NamespacedName{
+		Name:      obj.GetName(),
+		Namespace: obj.GetNamespace(),
+	}
+	return func() error {
+		return k.client.Get(k.context(), key, obj)
+	}
+}
+
+// List returns a function that
+func (k *komega) List(obj client.ObjectList, opts ...client.ListOption) func() error {
+	return func() error {
+		return k.client.List(k.context(), obj, opts...)
+	}
+}
+
+// Update tries to update an object by applying the updateFunc until the forwarded error matches.
+func (k *komega) Update(obj client.Object, updateFunc UpdateFunc, opts ...client.UpdateOption) func() error {
+	key := types.NamespacedName{
+		Name:      obj.GetName(),
+		Namespace: obj.GetNamespace(),
+	}
+	return func() error {
+		err := k.client.Get(k.context(), key, obj)
+		if err != nil {
+			return err
+		}
+		return k.client.Update(k.context(), updateFunc(obj), opts...)
+	}
+}
+
+// UpdateStatus tries to update an object's status by applying the updateFunc until the forwarded error matches.
+func (k *komega) UpdateStatus(obj client.Object, updateFunc UpdateFunc, opts ...client.UpdateOption) func() error {
+	key := types.NamespacedName{
+		Name:      obj.GetName(),
+		Namespace: obj.GetNamespace(),
+	}
+	return func() error {
+		err := k.client.Get(k.context(), key, obj)
+		if err != nil {
+			return err
+		}
+		return k.client.Status().Update(k.context(), updateFunc(obj), opts...)
+	}
+}
+
+// consistentlyclient.Object gets an individual object from the API server.
+func (k *komega) Object(obj client.Object) func() client.Object {
+	key := types.NamespacedName{
+		Name:      obj.GetName(),
+		Namespace: obj.GetNamespace(),
+	}
+	return func() client.Object {
+		err := k.client.Get(k.context(), key, obj)
+		if err != nil {
+			panic(err)
```

Comment:
Panicking here will cause the test to end on any error, which isn't what we want for an async assertion.
What you want to do instead is make this function return two values, the object and the error.

When Gomega sees two values, it will error if the second value is not nil/zero valued, so you can get away with returning errors and, if using this in an async assertion, it will wait until the error goes away. That means you can `Eventually(Object(&Doesn'tExistAtTheStart{})).ShouldNot(BeEmpty())`, or create the object and wait for it to be populated in the client cache if that makes sense.

So I would make this function:

```
func (k *komega) Object(obj client.Object) func() (client.Object, error) {
	key := types.NamespacedName{
		Name:      obj.GetName(),
		Namespace: obj.GetNamespace(),
	}
	return func() (client.Object, error) {
		err := k.client.Get(k.context(), key, obj)
		if err != nil {
			return nil, err
		}
		return obj, nil
	}
}
```
Same applies for the list below as well